homeassistant/components/upnp/config_flow.py

```diff
@@ -1,10 +1,172 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow start."""
+        _LOGGER.debug("async_step_user: user_input: %s", user_input)
+
+        if user_input is not None:
+            # Ensure wanted device was discovered.
+            matching_discoveries = [
+                discovery
+                for discovery in self._discoveries
+                if discovery["unique_id"] == user_input["unique_id"]
+            ]
+            if not matching_discoveries:
+                errors = {"base": "no_devices_discovered"}
+                return self.async_show_form(step_id="user", errors=errors,)
+
+            # Title/name will be updated later on.
+            self._data = matching_discoveries[0]
+            self._data[NAME] = urlparse(self._data[ssdp.ATTR_SSDP_LOCATION]).hostname
+            return self._async_create_entry_from_data()
+
+        # Discover devices.
+        self._discoveries = await Device.async_discover(self.hass)
+
+        # Filter discoveries which are already configured.
+        current_unique_ids = [
+            entry.unique_id for entry in self._async_current_entries()
+        ]
+        self._discoveries = [
+            discovery
+            for discovery in self._discoveries
+            if discovery["unique_id"] not in current_unique_ids
+        ]
+
+        # Ensure anything to add.
+        if not self._discoveries:
+            errors = {"base": "no_devices_discovered"}
+            return self.async_show_form(step_id="user", errors=errors,)
+
+        data_schema = vol.Schema(
+            {
+                vol.Required("unique_id"): vol.In(
+                    {
+                        discovery["unique_id"]: urlparse(
+                            discovery[ssdp.ATTR_SSDP_LOCATION]
+                        ).hostname
+                        for discovery in self._discoveries
+                    }
+                ),
+            }
+        )
+        return self.async_show_form(step_id="user", data_schema=data_schema,)
+
+    async def async_step_import(self, import_info: Optional[Mapping]):
+        """Import a new UPnP/IGD device as a config entry.
+
+        This flow is triggered by `async_setup`.
+        """
+        _LOGGER.debug("async_step_import: import_info: %s", import_info)
+
+        if import_info is None:
+            # Landed here via configuration.yaml entry.
+            # Any device already added, then abort.
+            if self._async_current_entries():
+                _LOGGER.debug("aborting, already configured")
+                return self.async_abort(reason="already_configured")
+
+        # Test if import_info isn't already configured.
+        if import_info is not None and any(
+            import_info["udn"] == entry.data["udn"]
+            and import_info["st"] == entry.data["st"]
+            for entry in self._async_current_entries()
+        ):
+            return self.async_abort(reason="already_configured")
```

Comment:
await self.async_set_unique_id(user_input['unique_id'], raise_on_progress=False)
self._abort_if_unique_id_configured({ stuff that should be updated on current config entry with same unique ID })